courseraprogramming/commands/oauth2.py

```diff
@@ -0,0 +1,493 @@
+
+        if self.local_webserver_port is not None:
+            # Boot up a local webserver to retrieve the response.
+            server_address = ('', self.local_webserver_port)
+            code_holder = CodeHolder()
+
+            local_server = BaseHTTPServer.HTTPServer(
+                server_address,
+                _make_handler(state_token, code_holder))
+
+            while not code_holder.has_code():
+                local_server.handle_request()
+            coursera_code = code_holder.code
+        else:
+            coursera_code = raw_input('Please enter the code from Coursera: ')
+
+        form_data = {
+            'code': coursera_code,
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'redirect_uri': self._redirect_uri,
+            'grant_type': 'authorization_code',
+        }
+        return self._request_tokens_from_token_endpoint(form_data)
+
+    def _exchange_refresh_tokens(self):
+        'Exchanges a refresh token for an access token'
+        if 'refresh' in self.token_cache:
+            # Attempt to use the refresh token to get a new access token.
+            refresh_form = {
+                'grant_type': 'refresh_token',
+                'refresh_token': self.token_cache['refresh'],
+                'client_id': self.client_id,
+                'client_secret': self.client_secret,
+            }
+            try:
+                tokens = self._request_tokens_from_token_endpoint(refresh_form)
+                tokens['refresh'] = self.token_cache['refresh']
+                return tokens
+            except OAuth2Exception:
+                logging.exception(
+                    'Encountered an exception during refresh token flow.')
+        return None
+
+    def _cache_has_good_token(self):
+        'The cache must have a token, and it must expire in > 5 minutes'
+        return (
+            self.token_cache is not None and
+            (time.time() + 5 * 60) < self.token_cache['expires']
+        )
+
+    def build_authorizer(self):
+        if not self._cache_has_good_token():
+            logging.debug('Attempting to use a refresh token.')
+            new_tokens = self._exchange_refresh_tokens()
+            if new_tokens is None:
+                logging.info(
+                    'Attempting to retrieve new tokens from the endpoint. You '
+                    'will be prompted to authorize the courseraprogramming '
+                    'app in your web browser.')
+                new_tokens = self._authorize_new_tokens()
+            logging.debug('New tokens: %s', new_tokens)
+            self.token_cache = new_tokens
+        else:
+            logging.debug('Local cache is good.')
+        return CourseraOAuth2Auth(self.token_cache['token'],
+                                  self.token_cache['expires'])
+
+
+def build_oauth2(args, cfg=None):
+    if cfg is None:
+        cfg = configuration()
+
+    try:
```

Comment:
I am using the [EAFP](https://docs.python.org/2/glossary.html#term-eafp) pattern, which is apparently the preferred approach.